[SPARK-10372] [CORE] basic test framework for entire spark scheduler

squito opened the pull request:
This is a basic framework for testing the entire scheduler.  The tests this adds aren't very interesting -- the point of this PR is just to setup the framework, to keep the initial change small, but it can be built upon to test more features (eg., speculation, killing tasks, blacklisting, etc.).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #41877 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41877/consoleFull) for   PR 8559 at commit [`844bd33`](https://github.com/apache/spark/commit/844bd339fa1913a5a42bf4efbb7bfffa9ba6b14b).

SparkQA commented:
[Test build #41877 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41877/console) for   PR 8559 at commit [`844bd33`](https://github.com/apache/spark/commit/844bd339fa1913a5a42bf4efbb7bfffa9ba6b14b).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41877/
Test FAILed.

squito commented:
Jenkins, retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #41878 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41878/consoleFull) for   PR 8559 at commit [`844bd33`](https://github.com/apache/spark/commit/844bd339fa1913a5a42bf4efbb7bfffa9ba6b14b).

squito commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -1123,8 +1133,15 @@ class DAGScheduler(
               s"$failedStage (${failedStage.name}) due to fetch failure")
+            // We might get lots of fetch failed for this stage, from lots of executors.
+            // Its better if we can resubmit for all the failed executors at one time, so lets
+            // just wait a *bit* before we resubmit.
```
I have no idea if this comment is accurate or not -- I just found this pretty confusing and felt it deserved some comment, so I took my best guess. The tests in `DAGSchedulerSuite` pass if you just post this event directly (with other corresponding changes to go along with it ...)

squito replied:
see https://github.com/apache/spark/pull/8560 to show that you can remove the messageScheduler completely (the tests pass, anyway ...)

squito commented on core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:
```diff
@@ -1242,7 +1259,7 @@ class DAGScheduler(
     } else {
       stage.latestInfo.stageFailed(errorMessage.get)
-      logInfo("%s (%s) failed in %s s".format(stage, stage.name, serviceTime))
+      logInfo(s"$stage (${stage.name}) failed in $serviceTime s due to ${errorMessage.get}")
```
the change here is adding `errorMessage`, the rest is just switching to using string interpolation

SparkQA commented:
[Test build #41878 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41878/console) for   PR 8559 at commit [`844bd33`](https://github.com/apache/spark/commit/844bd339fa1913a5a42bf4efbb7bfffa9ba6b14b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/41878/
Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #43989 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43989/consoleFull)** for PR 8559 at commit [`093a643`](https://github.com/apache/spark/commit/093a643f646552118cd17d2db4bbf2440f5723c1).

SparkQA commented:
**[Test build #43989 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43989/consoleFull)** for PR 8559 at commit [`093a643`](https://github.com/apache/spark/commit/093a643f646552118cd17d2db4bbf2440f5723c1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43989/
Test PASSed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45102 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45102/consoleFull)** for PR 8559 at commit [`609b61e`](https://github.com/apache/spark/commit/609b61eb32e482cbf843418fc7f921badf3bdabe).

SparkQA commented:
**[Test build #45102 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45102/consoleFull)** for PR 8559 at commit [`609b61e`](https://github.com/apache/spark/commit/609b61eb32e482cbf843418fc7f921badf3bdabe).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45102/
Test PASSed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45448 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45448/consoleFull)** for PR 8559 at commit [`609b61e`](https://github.com/apache/spark/commit/609b61eb32e482cbf843418fc7f921badf3bdabe).

SparkQA commented:
**[Test build #45448 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45448/consoleFull)** for PR 8559 at commit [`609b61e`](https://github.com/apache/spark/commit/609b61eb32e482cbf843418fc7f921badf3bdabe).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/45448/
Test PASSed.

SparkQA commented:
**[Test build #47164 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47164/consoleFull)** for PR 8559 at commit [`d61b13b`](https://github.com/apache/spark/commit/d61b13bf240b19c3db7252469f4189a7af3d95d4).

SparkQA commented:
**[Test build #47164 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47164/consoleFull)** for PR 8559 at commit [`d61b13b`](https://github.com/apache/spark/commit/d61b13bf240b19c3db7252469f4189a7af3d95d4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47164/
Test PASSed.

squito committed:
basic test framework for entire spark scheduler

SparkQA commented:
**[Test build #58266 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58266/consoleFull)** for PR 8559 at commit [`50d9162`](https://github.com/apache/spark/commit/50d9162e4638dfff7c4a5a5eb6c655c8e1fc9b58).

squito committed:
TaskResultGetter now expects there to always be non-null accum updates

SparkQA commented:
**[Test build #58275 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58275/consoleFull)** for PR 8559 at commit [`46ebdb6`](https://github.com/apache/spark/commit/46ebdb6cc77f5002d22b7f2fe34302b3fb381991).

SparkQA commented:
**[Test build #58266 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58266/consoleFull)** for PR 8559 at commit [`50d9162`](https://github.com/apache/spark/commit/50d9162e4638dfff7c4a5a5eb6c655c8e1fc9b58).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58266/
Test FAILed.

SparkQA commented:
**[Test build #58275 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58275/consoleFull)** for PR 8559 at commit [`46ebdb6`](https://github.com/apache/spark/commit/46ebdb6cc77f5002d22b7f2fe34302b3fb381991).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58275/
Test PASSed.

squito committed:
switch to making backend run in another thread

SparkQA commented:
**[Test build #58590 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58590/consoleFull)** for PR 8559 at commit [`6e8562f`](https://github.com/apache/spark/commit/6e8562fd152b76713fd64e128e69d5ded6fa4dbb).

SparkQA commented:
**[Test build #58590 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58590/consoleFull)** for PR 8559 at commit [`6e8562f`](https://github.com/apache/spark/commit/6e8562fd152b76713fd64e128e69d5ded6fa4dbb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class BasicSchedulerIntegrationSuite extends SchedulerIntegrationSuite[SingleCoreMockBackend] `
  * `class MultiExecutorBackend(`
  * `class ExecutorTaskStatus(val host: String, val executorId: String, var freeCores: Int)`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58590/
Test PASSed.

squito committed:
remove MultiExecutorBackend for now

SparkQA commented:
**[Test build #58593 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58593/consoleFull)** for PR 8559 at commit [`0b6da39`](https://github.com/apache/spark/commit/0b6da3909b45784fb2fc5abc644bac348fd1c6e8).

SparkQA commented:
**[Test build #58593 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58593/consoleFull)** for PR 8559 at commit [`0b6da39`](https://github.com/apache/spark/commit/0b6da3909b45784fb2fc5abc644bac348fd1c6e8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58593/
Test PASSed.

squito committed:
remove uncertain comment about messageScheduler

SparkQA commented:
**[Test build #58704 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58704/consoleFull)** for PR 8559 at commit [`c091187`](https://github.com/apache/spark/commit/c0911874783bdfadd7749a66a935cba2669f4ffa).

SparkQA commented:
**[Test build #58704 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58704/consoleFull)** for PR 8559 at commit [`c091187`](https://github.com/apache/spark/commit/c0911874783bdfadd7749a66a935cba2669f4ffa).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58704/
Test FAILed.

squito committed:
cleanup

SparkQA commented:
**[Test build #58712 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58712/consoleFull)** for PR 8559 at commit [`3b67b2a`](https://github.com/apache/spark/commit/3b67b2a950d75bc7f532c7a1151aaee864dc541f).

squito commented:
@mateiz @markhamstra @kayousterhout I've updated this to avoid changing the scheduler at all, and instead just run the mock backend in another thread.  It slightly complicates writing the tests, but avoids a (small) race in the tests and was probably necessary when we got to more complicated tests in any case.

In particular I want this for more tests on blacklisting ([SPARK-8426](https://github.com/apache/spark/pull/8760)), but I think there are a handful of open scheduler issues which will benefit from it.

SparkQA commented:
**[Test build #58712 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58712/consoleFull)** for PR 8559 at commit [`3b67b2a`](https://github.com/apache/spark/commit/3b67b2a950d75bc7f532c7a1151aaee864dc541f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58712/
Test PASSed.

squito committed:
add BlacklistIntegrationSuite and corresponding refactoring

SparkQA commented:
**[Test build #58785 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58785/consoleFull)** for PR 8559 at commit [`79bc384`](https://github.com/apache/spark/commit/79bc38416a914cebb33beaef3ec3179528848bca).

squito committed:
cleanup

squito commented:
sorry one more update -- I realized that some of the abstractions were too complicated for the simple tests I had, they were a step towards some more complex tests involving multiple executors.  I decided to just go ahead and add an example of those tests as well (which actually required some other minor changes in any case).

I'm sure there will be more to add to the framework at some point, but I think this is a good start for letting us add more tests.

SparkQA commented:
**[Test build #58786 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58786/consoleFull)** for PR 8559 at commit [`8349b76`](https://github.com/apache/spark/commit/8349b76ada807a9ce351cebbf4eddb88f67ca138).

squito committed:
comments

SparkQA commented:
**[Test build #58786 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58786/consoleFull)** for PR 8559 at commit [`8349b76`](https://github.com/apache/spark/commit/8349b76ada807a9ce351cebbf4eddb88f67ca138).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58786/
Test FAILed.

SparkQA commented:
**[Test build #58787 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58787/consoleFull)** for PR 8559 at commit [`7050b49`](https://github.com/apache/spark/commit/7050b49b4dfd6e7b4e3e966e9e15e3558a343d67).

squito committed:
move dummy killTask to MockBackend, otherwise occasional problems even in SingleCoreMockBackend when killTask is unsupported

SparkQA commented:
**[Test build #58788 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58788/consoleFull)** for PR 8559 at commit [`0095376`](https://github.com/apache/spark/commit/00953764f8d6a3807329ddff028c64a913cba662).

SparkQA commented:
**[Test build #58785 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58785/consoleFull)** for PR 8559 at commit [`79bc384`](https://github.com/apache/spark/commit/79bc38416a914cebb33beaef3ec3179528848bca).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class BlacklistIntegrationSuite extends SchedulerIntegrationSuite[MultiExecutorMockBackend]`
  * `class MultiExecutorMockBackend(`
  * `class MockRDDWithLocalityPrefs(`
  * `case class ExecutorTaskStatus(host: String, executorId: String, var freeCores: Int)`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58785/
Test PASSed.

squito committed:
take advantage of ExternalClusteManager extension

SparkQA commented:
**[Test build #58797 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58797/consoleFull)** for PR 8559 at commit [`8034995`](https://github.com/apache/spark/commit/8034995249e16e80fa7db5e709c38d1444b98f08).

squito committed:
cleanup

squito commented:
sorry one last update!  I just noticed that there was no need to change `SparkContext` -- what I needed was already added in https://issues.apache.org/jira/browse/SPARK-13904

SparkQA commented:
**[Test build #58799 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58799/consoleFull)** for PR 8559 at commit [`360c7cd`](https://github.com/apache/spark/commit/360c7cdf7731fd173bbfaf5e66d413f743604e60).

SparkQA commented:
**[Test build #58787 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58787/consoleFull)** for PR 8559 at commit [`7050b49`](https://github.com/apache/spark/commit/7050b49b4dfd6e7b4e3e966e9e15e3558a343d67).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58787/
Test PASSed.

SparkQA commented:
**[Test build #58788 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58788/consoleFull)** for PR 8559 at commit [`0095376`](https://github.com/apache/spark/commit/00953764f8d6a3807329ddff028c64a913cba662).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58788/
Test PASSed.

SparkQA commented:
**[Test build #58799 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58799/consoleFull)** for PR 8559 at commit [`360c7cd`](https://github.com/apache/spark/commit/360c7cdf7731fd173bbfaf5e66d413f743604e60).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58799/
Test FAILed.

SparkQA commented:
**[Test build #58797 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/58797/consoleFull)** for PR 8559 at commit [`8034995`](https://github.com/apache/spark/commit/8034995249e16e80fa7db5e709c38d1444b98f08).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/58797/
Test PASSed.

squito committed:
simplification and comments

squito commented:
last failure is unrelated, but pushed some minor updates in any case (not necessary for the tests here but helpful for other tests I'm adding)

SparkQA commented:
**[Test build #59035 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59035/consoleFull)** for PR 8559 at commit [`67acce9`](https://github.com/apache/spark/commit/67acce9a56a8acc24c0f2cf2ca76378277ec24d3).

SparkQA commented:
**[Test build #59035 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59035/consoleFull)** for PR 8559 at commit [`67acce9`](https://github.com/apache/spark/commit/67acce9a56a8acc24c0f2cf2ca76378277ec24d3).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59035/
Test FAILed.

squito commented:
Jenkins, retest this please

SparkQA commented:
**[Test build #59038 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59038/consoleFull)** for PR 8559 at commit [`67acce9`](https://github.com/apache/spark/commit/67acce9a56a8acc24c0f2cf2ca76378277ec24d3).

SparkQA commented:
**[Test build #59038 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59038/consoleFull)** for PR 8559 at commit [`67acce9`](https://github.com/apache/spark/commit/67acce9a56a8acc24c0f2cf2ca76378277ec24d3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59038/
Test PASSed.

markhamstra commented on core/src/main/scala/org/apache/spark/scheduler/TaskResult.scala:
```diff
@@ -59,7 +59,7 @@ private[spark] class DirectTaskResult[T](
     val numUpdates = in.readInt
     if (numUpdates == 0) {
-      accumUpdates = null
+      accumUpdates = Seq()
```
I'm a little surprised that we were getting away with that, given the number of places that are using DirectTaskResult#accumUpdates without checking for null.

squito replied:
yeah, likewise.  I guess in all real spark jobs, the number of accum updates is > 0, which is why this didn't particularly matter except for these mocks.  OTOH if this has any real danger we should just fix this separately (and be sure to include it in 2.0)

markhamstra commented on core/src/test/resources/META-INF/services/org.apache.spark.scheduler.ExternalClusterManager:
```diff
@@ -1 +1,2 @@
-org.apache.spark.scheduler.DummyExternalClusterManager
\ No newline at end of file
+org.apache.spark.scheduler.DummyExternalClusterManager
+org.apache.spark.scheduler.MockExternalClusterManager
```
Did you ever look at combining DummyExternalClusterManager and MockExternalClusterManager?  They are just two variations on a fake ExternalClusterManager for use in tests.  I realize that the focus of the tests for Dummy... and Mock... are different, so the two variations may not be easy or clean to combine, but if we could have just one fake ExternalClusterManager that still had a relatively clean implementation, I think that would be better than maintaining two.  OTOH, if combining them gets messy, then just go with what you've already got.

squito replied:
I did not, mostly because I thought Dummy should be left as is, to keep ExternalClusterManagerSuite cleanly focused.  It would be easy to combine them, though.  Dummy isn't covering any cases that Mock isn't -- I could remove Dummy, and make ExternalClusterManagerSuite reference Mock, without making any changes to Mock.

SparkQA commented:
**[Test build #59130 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59130/consoleFull)** for PR 8559 at commit [`1fdf2aa`](https://github.com/apache/spark/commit/1fdf2aabc7b18ca68399d1a8e586ce41ed5d2bb9).

squito commented:
@markhamstra thanks for the review.  I went ahead and added one more commit and removed `DummyExternalClusterManager`, just used Mock in the old tests.  Honestly I kinda prefer having `ExternalClusterManagerSuite` being very simple and self-contained, but also don't feel particularly strongly about it.  (easy to back out since its just one final commit.)

markhamstra commented:
@squito I don't feel really strongly about it, either.  My only concern was for others adding tests needing a mock ExternalClusterManager in the future and not knowing which one to use, why they are different, whether any new mocking needs to be added to both, etc.

If someone does feel strongly about maintaining the separation, then we can put things back the way you had them.

SparkQA commented:
**[Test build #59130 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59130/consoleFull)** for PR 8559 at commit [`1fdf2aa`](https://github.com/apache/spark/commit/1fdf2aabc7b18ca68399d1a8e586ce41ed5d2bb9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59130/
Test PASSed.

squito commented:
@markhamstra alright, in that case do you have any objections if I merge this, one commit back?  I'll throw in a comment on DummyExternalClusterManager pointing to MockExternalClusterManager, as I think most folks writing tests would rather start from there.

markhamstra commented:
@squito Yeah, that's fine.  I haven't gone through the new tests closely to make sure that they are doing what they say they are doing, but the changes to both non-test code and previous tests look safe.

squito committed:
check initialization and add comments

SparkQA commented:
**[Test build #59226 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59226/consoleFull)** for PR 8559 at commit [`278122e`](https://github.com/apache/spark/commit/278122e30c1cf9386d835e204038e141e900bea6).

SparkQA commented:
**[Test build #59226 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59226/consoleFull)** for PR 8559 at commit [`278122e`](https://github.com/apache/spark/commit/278122e30c1cf9386d835e204038e141e900bea6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59226/
Test PASSed.

rxin commented:
This is pretty cool!

rxin commented:
on a related note, @squito can you in the future leave a msg indicating the branch a pr was merged once you merge it? There have been cases that lead to race conditions in merging and also mistakes in the branches that we needed to go back and audit.

squito commented:
sure thing, sorry about that @rxin !  fwiw, this was just merged to master.

rxin commented:
np - a lot of people forget to do this. I am just bugging everybody :)

rxin commented:
btw i've seen a few flaky tests introduced by this one. If we see more of them we might need to disable and increase timeouts.

squito commented:
Oh thanks for pointing out the flaky test, I think I figured out the problem, opened https://issues.apache.org/jira/browse/SPARK-15714.  Weird I never this running locally -- I'd expect this race to be pretty common.  Just tried and an on my laptop it only shows up once in 100 runs.

skonto commented:
A test is failing:

`Map(8 -> 42, 7 -> 42) was not empty
ScalaTestFailureLocation: org.apache.spark.scheduler.BlacklistIntegrationSuite$$anonfun$1 at (BlacklistIntegrationSuite.scala:51)
org.scalatest.exceptions.TestFailedException: Map(8 -> 42, 7 -> 42) was not empty
	at org.scalatest.Assertions$class.newAssertionFailedException(Assertions.scala:500)
	at org.scalatest.FunSuite.newAssertionFailedException(FunSuite.scala:1555)
	at org.scalatest.Assertions$AssertionsHelper.macroAssert(Assertions.scala:466)
	at org.apache.spark.scheduler.BlacklistIntegrationSuite$$anonfun$1.apply$mcV$sp(BlacklistIntegrationSuite.scala:51)
	at org.apache.spark.scheduler.SchedulerIntegrationSuite$$anonfun$testScheduler$1.apply$mcV$sp(SchedulerIntegrationSuite.scala:88)
	at org.apache.spark.scheduler.SchedulerIntegrationSuite$$anonfun$testScheduler$1.apply(SchedulerIntegrationSuite.scala:84)
	at org.apache.spark.scheduler.SchedulerIntegrationSuite$$anonfun$testScheduler$1.apply(SchedulerIntegrationSuite.scala:84)
	at org.scalatest.Transformer$$anonfun$apply$1.apply$mcV$sp(Transformer.scala:22)
	at org.scalatest.OutcomeOf$class.outcomeOf(OutcomeOf.scala:85)
	at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
	at org.scalatest.Transformer.apply(Transformer.scala:22)
	at org.scalatest.Transformer.apply(Transformer.scala:20)
	at org.scalatest.FunSuiteLike$$anon$1.apply(FunSuiteLike.scala:166)
	at org.apache.spark.SparkFunSuite.withFixture(SparkFunSuite.scala:57)
	at org.scalatest.FunSuiteLike$class.invokeWithFixture$1(FunSuiteLike.scala:163)
	at org.scalatest.FunSuiteLike$$anonfun$runTest$1.apply(FunSuiteLike.scala:175)
	at org.scalatest.FunSuiteLike$$anonfun$runTest$1.apply(FunSuiteLike.scala:175)
	at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
	at org.scalatest.FunSuiteLike$class.runTest(FunSuiteLike.scala:175)
	at org.apache.spark.scheduler.SchedulerIntegrationSuite.org$scalatest$BeforeAndAfterEach$$super$runTest(SchedulerIntegrationSuite.scala:44)
	at org.scalatest.BeforeAndAfterEach$class.runTest(BeforeAndAfterEach.scala:255)
	at org.apache.spark.scheduler.SchedulerIntegrationSuite.runTest(SchedulerIntegrationSuite.scala:44)
	at org.scalatest.FunSuiteLike$$anonfun$runTests$1.apply(FunSuiteLike.scala:208)
	at org.scalatest.FunSuiteLike$$anonfun$runTests$1.apply(FunSuiteLike.scala:208)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:413)
	at org.scalatest.SuperEngine$$anonfun$traverseSubNodes$1$1.apply(Engine.scala:401)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at org.scalatest.SuperEngine.traverseSubNodes$1(Engine.scala:401)
	at org.scalatest.SuperEngine.org$scalatest$SuperEngine$$runTestsInBranch(Engine.scala:396)
	at org.scalatest.SuperEngine.runTestsImpl(Engine.scala:483)
	at org.scalatest.FunSuiteLike$class.runTests(FunSuiteLike.scala:208)
	at org.scalatest.FunSuite.runTests(FunSuite.scala:1555)
	at org.scalatest.Suite$class.run(Suite.scala:1424)
	at org.scalatest.FunSuite.org$scalatest$FunSuiteLike$$super$run(FunSuite.scala:1555)
	at org.scalatest.FunSuiteLike$$anonfun$run$1.apply(FunSuiteLike.scala:212)
	at org.scalatest.FunSuiteLike$$anonfun$run$1.apply(FunSuiteLike.scala:212)
	at org.scalatest.SuperEngine.runImpl(Engine.scala:545)
	at org.scalatest.FunSuiteLike$class.run(FunSuiteLike.scala:212)
	at org.apache.spark.SparkFunSuite.org$scalatest$BeforeAndAfterAll$$super$run(SparkFunSuite.scala:29)
	at org.scalatest.BeforeAndAfterAll$class.liftedTree1$1(BeforeAndAfterAll.scala:257)
	at org.scalatest.BeforeAndAfterAll$class.run(BeforeAndAfterAll.scala:256)
	at org.apache.spark.SparkFunSuite.run(SparkFunSuite.scala:29)
	at org.scalatest.tools.SuiteRunner.run(SuiteRunner.scala:55)
	at org.scalatest.tools.Runner$$anonfun$doRunRunRunDaDoRunRun$3.apply(Runner.scala:2563)
	at org.scalatest.tools.Runner$$anonfun$doRunRunRunDaDoRunRun$3.apply(Runner.scala:2557)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at org.scalatest.tools.Runner$.doRunRunRunDaDoRunRun(Runner.scala:2557)
	at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1044)
	at org.scalatest.tools.Runner$$anonfun$runOptionallyWithPassFailReporter$2.apply(Runner.scala:1043)
	at org.scalatest.tools.Runner$.withClassLoaderAndDispatchReporter(Runner.scala:2722)
	at org.scalatest.tools.Runner$.runOptionallyWithPassFailReporter(Runner.scala:1043)
	at org.scalatest.tools.Runner$.run(Runner.scala:883)
	at org.scalatest.tools.Runner.run(Runner.scala)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.runScalaTest2(ScalaTestRunner.java:138)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.main(ScalaTestRunner.java:28)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:144)`

Also check here:
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59928/

squito commented:
@skonto I just merged https://github.com/apache/spark/pull/13454, please let me know if that does not address the issue.

skonto commented:
great thnx, it passes locally now. I will test my pr too.

squito commented:
@skonto since you seem to be able to reproduce this regularly though I can't -- can you try just increasing the timeouts?  honestly I'm surprised you're hitting them, but if we just need to bump up the timeouts a little more, we can do that.  If there is something else wrong, eg. another race, than lets turn these tests off for now till I can nail it down.

skonto commented:
@squito not really check here: 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/60037/

Its not only my issue and i dont think increasing timeouts is a solution. Tests should be stable.

squito commented:
@skonto agreed, I have opened https://issues.apache.org/jira/browse/SPARK-15783 + PR https://github.com/apache/spark/pull/13528 to turn them off for now.

I was at least able to reproduce the failures with 5k runs on my laptop, (also seems worse when I run all the different tests together vs just running one test repeatedly) so I will try to nail down the cause more before adding back.

squito commented:
@skonto I just merged https://github.com/apache/spark/pull/13528 this to turn the tests off for now.  Sorry again about the issues, and I'll work on figuring out the cause of the flakiness
